### PR TITLE
Add extra attributes to Link and use for HAL mediatype

### PR DESCRIPTION
**Pull request by gregturn**
Adds many additional attributes defined in RFC5988 and verifies they work properly in the neutral representation of Link while also being rendered properly in the HAL module.

Related tickets: #100, #417, #235
Previous pull requests: #240, #238, #223, #79

### CODE REVIEW

**Comment by gregturn** [comment]
@olivergierke I redid this PR so that it doesn't attempt to clean up any whitespace.

**Comment by netwolfuk** [comment]
Hi @gregturn @olivergierke 
I'm just wondering if you have an ETA for 0.24 release?
Many thanks.

**Comment by gregturn** [comment]
@netwolfuk I can certainly comment that Oliver was out for a couple weeks, so he has been catching up on a stack of stuff across the Spring Data suite, the biggest HATEOAS one being the [Affordances API effort](https://github.com/spring-projects/spring-hateoas/pull/581). The goal is to at least have that pulled in. 

I'm working to get several other patches pulled in as well (may be at least 7 other PRs I have yet to review with Oliver).

What may be highly pertinent to this PR is that the Affordances PR ALSO pulls in extra RFC 5988 material, so I'll have to certainly reevaluate this one when the dust settles.

As to how many fit into 0.24, that has yet to be determined. It may be shaped by the state of Spring Data's release train when we finish Affordances.

Sorry about being so vague. Wanted to give you insight into the process.

**Comment by netwolfuk** [comment]
@gregturn Thanks for the thorough response. I really appreciate all you guys do.

**Comment by nstaeger** [comment]
Would highly appreciate if this would be merged, as this would be very useful to us. Thanks for implementing @gregturn

**Comment by gregturn** [comment]
I'm shooting to get all the RFC 5988 material included. Just want a coherent solution between this PR and Affordances.

File: src/main/java/org/springframework/hateoas/Link.java [hunk 1]
```diff
@@ -94,6 +103,27 @@ public Link(UriTemplate template, String rel) {
...
+	 *
+	 * @param href must not be {@literal null} or empty.
+	 * @param rel must not be {@literal null} or empty.
+	 * @param hreflang
+	 * @param media
+	 * @param title
+	 * @param type
+	 * @param deprecation
+	 */
+	public Link(String href, String rel, String hreflang, String media, String title, String type, String deprecation) {
```
Comment (olivergierke) [inline src/main/java/org/springframework/hateoas/Link.java]: Does it make sense to use Lombok here?

File: src/main/java/org/springframework/hateoas/Link.java [hunk 2]
```diff
@@ -119,6 +149,51 @@ public String getRel() {
 	}
 
 	/**
+	 * Returns the hreflang of the link.
+	 *
+	 * @return
+	 */
+	public String getHreflang() {
```
Comment (olivergierke) [inline src/main/java/org/springframework/hateoas/Link.java]: Lombok?

File: src/main/java/org/springframework/hateoas/Link.java [hunk 3]
```diff
@@ -138,6 +213,66 @@ public Link withSelfRel() {
 	}
 
 	/**
+	 * Returns a {@link Link} with the {@code hreflang} attribute filled out.
+	 *
+	 * @param hreflang
+	 * @return
+	 */
+	public Link withHreflang(String hreflang) {
```
Comment (olivergierke) [inline src/main/java/org/springframework/hateoas/Link.java]: Lombok's `@Wither`? Should we have `withHref(…)` and `withRel(…)`, too, for consistency reasons?

File: src/main/java/org/springframework/hateoas/Link.java [hunk 4]
```diff
@@ -212,7 +347,20 @@ public boolean equals(Object obj) {
 
 		Link that = (Link) obj;
 
-		return this.href.equals(that.href) && this.rel.equals(that.rel);
```
Comment (olivergierke) [inline src/main/java/org/springframework/hateoas/Link.java]: `@EqualsAndHashCode`?

File: src/main/java/org/springframework/hateoas/Link.java [hunk 5]
```diff
@@ -263,7 +448,34 @@ public static Link valueOf(String element) {
 				throw new IllegalArgumentException("Link does not provide a rel attribute!");
 			}
 
-			return new Link(matcher.group(1), attributes.get("rel"));
+			Set<String> unrecognizedHeaders = unrecognizedHeaders(attributes);
```
Comment (olivergierke) [inline src/main/java/org/springframework/hateoas/Link.java]: Following Postel's Law I think we should not throw an exception here but just ignore attributes that we don't understand.

File: src/main/java/org/springframework/hateoas/Link.java (hunk 1 above)
Comment (gregturn) [inline src/main/java/org/springframework/hateoas/Link.java]: Done

File: src/main/java/org/springframework/hateoas/Link.java (hunk 2 above)
Comment (gregturn) [inline src/main/java/org/springframework/hateoas/Link.java]: Done

File: src/main/java/org/springframework/hateoas/Link.java (hunk 3 above)
Comment (gregturn) [inline src/main/java/org/springframework/hateoas/Link.java]: Done

File: src/main/java/org/springframework/hateoas/Link.java (hunk 4 above)
Comment (gregturn) [inline src/main/java/org/springframework/hateoas/Link.java]: Done

File: src/main/java/org/springframework/hateoas/Link.java (hunk 5 above)
Comment (gregturn) [inline src/main/java/org/springframework/hateoas/Link.java]: Done

**Comment by olivergierke** [comment]
That's merged.

**Comment by VelDeveloper** [comment]
@olivergierke @gregturn I am facing the similar issue... I would like to add an extra attribute along with "href". May I know how to do it? Any help would be appreciable. I am using spring-boot 1.5.8 along with Hateoas

**Pull request closed without merging**